File: 0. prereq/typescript-test/enroll-q3.ts
import { AnchorProvider, Program, Wallet } from "@coral-xyz/anchor";
import { Connection, Keypair, PublicKey, SystemProgram } from "@solana/web3.js";
import * as fs from "fs";

const secret = JSON.parse(fs.readFileSync("./dev-wallet.json", "utf-8"));
const keypair = Keypair.fromSecretKey(new Uint8Array(secret));

// Set up the connection to Solana Devnet
const connection = new Connection("https://api.devnet.solana.com", "confirmed");

// Create a wallet and provider
const wallet = new Wallet(keypair);
const provider = new AnchorProvider(connection, wallet, { commitment: "confirmed" });

// Program ID from the provided address
const programId = new PublicKey("TRBZyQHB3m68FGeVsqTK39Wm4xejadjVhP5MAZaKWDM");

// Your github username
const github = "rxrxrxrx"; // Replace with your actual github username

// Collection address from Q3 subject
const mintCollection = new PublicKey("5ebsp5RChCGK7ssRZMVMufgVZhd2kFbNaotcZ5UvytN2");

// MPL Core Program ID
const MPL_CORE_PROGRAM_ID = new PublicKey("CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d");

// Generate mint keypair for the NFT
const mintTs = Keypair.generate();

// Main function
(async () => {
  try {
    // Fetch the IDL directly from the blockchain
    const idl = await Program.fetchIdl(programId, provider);
    if (!idl) {
      throw new Error("IDL not found for the program");
    }
    
    console.log("IDL fetched successfully");
    
    // Create the program with the fetched IDL
    const program = new Program(idl, provider);
    
    // Create the PDA for the account with correct seed "prereqs"
    const account_seeds = [
      Buffer.from("prereqs"),
      keypair.publicKey.toBuffer(),
    ];
    
    const [account_key, _account_bump] = PublicKey.findProgramAddressSync(
      account_seeds,
      programId
    );
    
    // Create the authority PDA for the collection
    const authority_seeds = [
      Buffer.from("collection"),
      mintCollection.toBuffer(),
    ];
    
    const [authority, _authority_bump] = PublicKey.findProgramAddressSync(
      authority_seeds,
      programId
    );
    
    console.log("Account PDA address:", account_key.toString());
    console.log("Authority PDA address:", authority.toString());
    console.log("Mint address:", mintTs.publicKey.toString());
    
    // Execute the initialize transaction
    console.log("\n=== Executing Initialize Transaction ===");
    try {
      const txhash1 = await program.methods
        .initialize(github)
        .accountsPartial({
          user: keypair.publicKey,
          account: account_key,
          system_program: SystemProgram.programId,
        })
        .signers([keypair])
        .rpc();
        
      console.log(`Success:
      https://explorer.solana.com/tx/${txhash1}?cluster=devnet`);
    } catch (e) {
      console.error(`Initialize failed: ${e}`);
      return;
    }
    
    // Wait a bit between transactions
    await new Promise(resolve => setTimeout(resolve, 2000));
    
    // Execute the submit_ts transaction
    console.log("\n=== Executing Submit TS Transaction ===");
    try {
      const txhash2 = await program.methods
        .submitTs()
        .accountsPartial({
          user: keypair.publicKey,
          account: account_key,
          mint: mintTs.publicKey,
          collection: mintCollection,
          authority: authority,
          mpl_core_program: MPL_CORE_PROGRAM_ID,
          system_program: SystemProgram.programId,
        })
        .signers([keypair, mintTs])
        .rpc();
        
      console.log(`Submit:
      https://explorer.solana.com/tx/${txhash2}?cluster=devnet`);
      
      console.log(`NFT minted at: ${mintTs.publicKey.toString()}`);
      
    } catch (e) {
      console.error(`Submit TS failed: ${e}`);
    }
    
  } catch(e) {
    console.error(`Something went wrong: ${e}`);
  }
})(); 

File: 0. prereq/rust-test/src/enroll-q3.rs
use solana_client::rpc_client::RpcClient;
use solana_sdk::{
    instruction::{AccountMeta, Instruction},
    pubkey::Pubkey,
    signature::{Keypair, Signer, read_keypair_file},
    system_program,
    transaction::Transaction,
};
use std::str::FromStr;

// Constants
const RPC_URL: &str = "https://api.devnet.solana.com";
const PROGRAM_ID: &str = "TRBZyQHB3m68FGeVsqTK39Wm4xejadjVhP5MAZaKWDM";
const COLLECTION_ADDRESS: &str = "5ebsp5RChCGK7ssRZMVMufgVZhd2kFbNaotcZ5UvytN2";
const MPL_CORE_PROGRAM_ID: &str = "CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d";

// PDA account for "prereqs"
fn derive_account_address(user: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[b"prereqs", user.as_ref()],
        &Pubkey::from_str(PROGRAM_ID).unwrap(),
    )
}

// PDA authority for "collection"
fn derive_authority_address(collection: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &[b"collection", collection.as_ref()],
        &Pubkey::from_str(PROGRAM_ID).unwrap(),
    )
}

// submit_rs instruction
fn create_submit_rs_instruction(
    user: &Pubkey,
    account: &Pubkey,
    mint: &Pubkey,
    collection: &Pubkey,
    authority: &Pubkey,
) -> Instruction {
    // discriminator for 'submit_rs' from IDL
    let data = vec![77, 124, 82, 163, 21, 133, 181, 206];

    Instruction {
        program_id: Pubkey::from_str(PROGRAM_ID).unwrap(),
        accounts: vec![
            AccountMeta::new(*user, true),              // user (writable + signer)
            AccountMeta::new(*account, false),          // account (writable)
            AccountMeta::new(*mint, true),              // mint (writable + signer)
            AccountMeta::new(*collection, false),       // collection (writable)
            AccountMeta::new_readonly(*authority, false), // authority
            AccountMeta::new_readonly(Pubkey::from_str(MPL_CORE_PROGRAM_ID).unwrap(), false), // mpl_core_program
            AccountMeta::new_readonly(system_program::id(), false), // system_program
        ],
        data,
    }
}

#[test]
fn enroll_q3() {
    let client = RpcClient::new(RPC_URL);
    
    let signer = read_keypair_file("Turbin3-wallet.json")
        .expect("Couldn't find wallet file");
    
    // PDA account for "prereqs"
    let (account, _bump) = derive_account_address(&signer.pubkey());
    
    // Collection address
    let collection = Pubkey::from_str(COLLECTION_ADDRESS).unwrap();

    // PDA authority for "collection"
    let (authority, _authority_bump) = derive_authority_address(&collection);
    
    // mint keypair
    let mint = Keypair::new();
    
    println!("User: {}", signer.pubkey());
    println!("Account PDA: {} (already initialized from TypeScript)", account);
    println!("Collection: {}", collection);
    println!("Authority PDA: {}", authority);
    println!("Mint: {}", mint.pubkey());
    
    println!("\n Executing Submit RS Transaction");
    
    let submit_rs_instruction = create_submit_rs_instruction(
        &signer.pubkey(),
        &account,
        &mint.pubkey(),
        &collection,
        &authority,
    );
    
    let recent_blockhash = client
        .get_latest_blockhash()
        .expect("Failed to get recent blockhash");
    
    let submit_rs_transaction = Transaction::new_signed_with_payer(
        &[submit_rs_instruction],
        Some(&signer.pubkey()),
        &[&signer, &mint], 
        recent_blockhash,
    );
    
    match client.send_and_confirm_transaction(&submit_rs_transaction) {
        Ok(signature) => {
            println!("Submit Success!");
            println!("https://explorer.solana.com/tx/{}?cluster=devnet", signature);
            println!("NFT minted at: {}", mint.pubkey());
        },
        Err(e) => {
            println!("Submit RS failed: {}", e);
        }
    }
} 